Making role permissions more granular and avoid using *. #4188

subash89 opened the pull request:
# Description
Dapr rbac enhancements as in #4188 

<!--
Please explain the changes you've made.
-->

## Issue reference
https://github.com/dapr/dapr/issues/4188
<!--
We strive to have all PR being opened based on an issue, where the problem or feature have been discussed prior to implementation.
-->

Please reference the issue this PR will close: #_[issue number]_

## Checklist

Please make sure you've  completed the relevant tasks for this PR, out of the following list:

* [ ] Code compiles correctly
* [ ] Created/updated tests
* [ ] Unit tests passing
* [ ] End-to-end tests passing
* [ ] Extended the documentation / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Specification has been updated / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_
* [ ] Provided sample for the feature / Created issue in the https://github.com/dapr/docs/ repo: dapr/docs#_[issue number]_

subash89 commented:
@yaron2  please refer this for the new PR for ClusterRoleBinding rbac changes.

subash89 commented:
I resolved the DCO conflict.

yaron2 commented on charts/dapr/charts/dapr_rbac/templates/ClusterRoleBinding.yaml:
```diff
@@ -29,46 +29,138 @@ apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: secret-reader
-  namespace: default
+  namespace: {{ .Release.Namespace }}
```
This actually needs to be in default.

subash89 replied:
Can you explain why this needs to be in default namespace ? 
For example:  My app+side car in namespace=myapp123 and dapr controlplane namesapce=dapr-system. Why we need this in default namespace. Is default is meant to be the one that the side car deployed in  ?(allowing side car to access secrets) ?

yaron2 replied:
These roles are intended for the Secrets API specifically, and this is why they are set up automatically on the `default` namespace which is where apps (deployments) are running by default on Kubernetes. This role is not intended for the Dapr control plane and thus should not be in the `Release.Namespace`.

yaron2 commented on charts/dapr/charts/dapr_rbac/templates/ClusterRoleBinding.yaml:
```diff
@@ -29,46 +29,138 @@ apiVersion: rbac.authorization.k8s.io/v1
 metadata:
   name: dapr-secret-reader
-  namespace: default
+  namespace:  {{ .Release.Namespace }}
```
This needs to be default too.

subash89 replied:
Same reasoning as my earlier comment .

yaron2 commented:
@subash89 thanks, I gave two small feedback comments. Please resolve the additional conflict, and if all tests pass, we can merge! :)

subash89 commented:
Is PR all good ? DCO and merge conflict seems fine.

yaron2 commented:
> Is PR all good ? DCO and merge conflict seems fine.

See my comment above. Also, all end to end tests are failing on installing Dapr on k8s, so something is wrong..

subash89 commented:
> > Is PR all good ? DCO and merge conflict seems fine.
> 
> See my comment above. Also, all end to end tests are failing on installing Dapr on k8s, so something is wrong..

@yaron2  the failure [1] happens due to the issue opened in https://github.com/dapr/dapr/issues/4321
I removed them in this PR as I opened a new issue for it.(#4321)

This likely working in dev branch, as the permission is given flat out * in ClusterRoleBinding.yml. After adding the fix I mentioned in the above issue, including the secret-reader and it's binding added into dapr namespace instead of default(or the secret reader namespace), the install went fine.

Can you tell me why secret reader has its own namespace and I did not see any use case so far from deployment point of view.

[1]
time="2022-03-14T20:52:46.159088283Z" level=error msg="error checking if credentials exist: secrets \"dapr-trust-bundle\" is forbidden: User \"system:serviceaccount:dapr-system:dapr-operator\" cannot get resource \"secrets\" in API group \"\" in the namespace \"dapr-system\"" instance=dapr-sentry-64f754bbfb-sqf2m scope=dapr.sentry.ca type=log ver=edge
time="2022-03-14T20:52:46.159140467Z" level=info msg="root and issuer certs not found: generating self signed CA" instance=dapr-sentry-64f754bbfb-sqf2m scope=dapr.sentry.ca type=log ver=edge
time="2022-03-14T20:52:46.163394956Z" level=fatal msg="error loading trust root bundle: error generating trust root bundle: failed saving secret to kubernetes: secrets \"dapr-trust-bundle\" is forbidden: User \"system:serviceaccount:dapr-system:dapr-operator\" cannot update resource \"secrets\" in API group \"\" in the namespace \"dapr-system\"" instance=dapr-sentry-64f754bbfb-sqf2m scope=dapr.sentry type=log ver=edge

yaron2 commented:
> > > Is PR all good ? DCO and merge conflict seems fine.
> > 
> > 
> > See my comment above. Also, all end to end tests are failing on installing Dapr on k8s, so something is wrong..
> 
> @yaron2 the failure [1] happens due to the issue opened in #4321 I removed them in this PR as I opened a new issue for it.(#4321)
> 
> This likely working in dev branch, as the permission is given flat out * in ClusterRoleBinding.yml. After adding the fix I mentioned in the above issue, including the secret-reader and it's binding added into dapr namespace instead of default(or the secret reader namespace), the install went fine.
> 
> Can you tell me why secret reader has its own namespace and I did not see any use case so far from deployment point of view.
> 
> [1] time="2022-03-14T20:52:46.159088283Z" level=error msg="error checking if credentials exist: secrets "dapr-trust-bundle" is forbidden: User "system:serviceaccount:dapr-system:dapr-operator" cannot get resource "secrets" in API group "" in the namespace "dapr-system"" instance=dapr-sentry-64f754bbfb-sqf2m scope=dapr.sentry.ca type=log ver=edge time="2022-03-14T20:52:46.159140467Z" level=info msg="root and issuer certs not found: generating self signed CA" instance=dapr-sentry-64f754bbfb-sqf2m scope=dapr.sentry.ca type=log ver=edge time="2022-03-14T20:52:46.163394956Z" level=fatal msg="error loading trust root bundle: error generating trust root bundle: failed saving secret to kubernetes: secrets "dapr-trust-bundle" is forbidden: User "system:serviceaccount:dapr-system:dapr-operator" cannot update resource "secrets" in API group "" in the namespace "dapr-system"" instance=dapr-sentry-64f754bbfb-sqf2m scope=dapr.sentry type=log ver=edge

These roles are intended for the Secrets API specifically, and this is why they are set up automatically on the default namespace which is where apps (deployments) are running by default on Kubernetes. This role is not intended for the Dapr control plane and thus should not be in the Release.Namespace.

subash89 commented:
Great.  Thanks for the clarification. 

I Like the secret namespace is now parameterized, to use secrets api from application teams. But ideally it could  be multiple namespaces as multiple teams likely be in their own namespaces. But in that case additional role bindings can be added separately.

I updated the PR with a new commit which I added secrets permission for dapr-operator. Now should be all good. I verified with a minikube,

subash89 commented:
Any update on this pls ?

yaron2 commented:
> Any update on this pls ?

Please resolve conflicts, and if the tests pass we can merge.

yaron2 commented:
> > Any update on this pls ?
> 
> Please resolve conflicts, and if the tests pass we can merge.

ping @subash89.

subash89 commented:
@yaron2  Added the changes.

subash89 commented:
The conflict this suggest is the actual diff of the PR. I assume this is false positive then ?

subash89 commented:
@yaron2  is this all good ?

yaron2 commented:
> @yaron2 is this all good ?

I resolved the conflict, waiting for the tests to pass now.

subash89 commented:
Looks like tests failed once again. I ran the tests/configs all deployement yamls againt a minikube. Is there a way I can run this triggered test locally ?

yaron2 commented:
/ok-to-test

yaron2 commented:
Ping @subash89

yaron2 commented:
@subash89 can you please address the missing resource so we can merge this?

subash89 commented:
I was trying to run this test against local minikube behind cooperate proxies and having a hard time. Let me add the missing resiliency.

subash89 commented:
Added the changes to the PR

yaron2 commented:
Thanks @subash89!
